### PR TITLE
Honor global timeout during snapshot save

### PR DESCRIPTION
**Pull request by tiwillia**
Bug 1255966
https://bugzilla.redhat.com/show_bug.cgi?id=1255966

When saving a snapshot, honor the timeout passed as a global argument, or defined in configuration.

### CODE REVIEW

**Comment by tiwillia** [comment]
[test] although I believe there is some missed coverage now due to splitting the `Timeout::Error` rescue from the rest of possible rescued errors.

File: lib/rhc/ssh_helpers.rb [hunk 1]
```diff
@@ -256,29 +257,43 @@ def save_snapshot(app, filename, for_deployment=false, ssh_executable=nil)
       ssh_stderr = " 2>/dev/null"
       debug ssh_cmd
 
+      # Default timeout is 30 minutes, more than enough time to ensure the snapshot is copied successfully.
+      options.timeout ? exec_timeout = options.timeout : exec_timeout = 1800
```
Comment (Miciah) [inline lib/rhc/ssh_helpers.rb]: This is unidiomatic.  How about the following?
```ruby
exec_timeout = options.timeout || 1800
```

File: lib/rhc/ssh_helpers.rb [hunk 2]
```diff
@@ -256,29 +257,43 @@ def save_snapshot(app, filename, for_deployment=false, ssh_executable=nil)
...
+
       say "Pulling down a snapshot of application '#{app.name}' to #{filename} ... "
 
       begin
         if !RHC::Helpers.windows?
+          status = 1
+          output = ""
+          Timeout::timeout(exec_timeout) {
             status, output = exec(ssh_cmd + (debug? ? '' : ssh_stderr))
```
Comment (Miciah) [inline lib/rhc/ssh_helpers.rb]: Hm, why don't we conditionally set `ssh_stderr = debug? '' : ' 2>/dev/null'` above and then unconditionally interpolate `exec(ssh_cmd + ssh_stderr)` here? But I guess that's outside of scope for this PR.

File: lib/rhc/ssh_helpers.rb [hunk 3]
```diff
@@ -256,29 +257,43 @@ def save_snapshot(app, filename, for_deployment=false, ssh_executable=nil)
...
+          end
         else
-          Net::SSH.start(ssh_uri.host, ssh_uri.user) do |ssh|
-            File.open(filename, 'wb') do |file|
-              ssh.exec! snapshot_cmd do |channel, stream, data|
-                if stream == :stdout
-                  file.write(data)
-                else
-                  debug data
+          Timeout::timeout(exec_timeout) {
```
Comment (Miciah) [inline lib/rhc/ssh_helpers.rb]: It's a minor thing, but it's generally preferable to use `do`/`end` instead of braces for long blocks like this one.

**Comment by thrasher-redhat** [comment]
Looks like the previous rescue/raise (ssh_helpers.rb old lines 281-283) were all being covered by the "when timing out on windows" test.  The new general timeout test case also covers that same rescue/raise.  That  old rescue/raise has now been split into the timeout error (covered) and other network related issues (no longer covered).

So it looks like there needs to be a new test that fails one (or all of) the exceptions listed on new line 296.

File: lib/rhc/ssh_helpers.rb (hunk 2 above)
Comment (tiwillia) [inline lib/rhc/ssh_helpers.rb]: I noticed this too, but figured this same - out of scope.

**Comment by tiwillia** [comment]
Should be passing coverage [test]s now

**Comment by Miciah** [comment]
[merge] please!

**Pull request merged**